### PR TITLE
Create 0009-versioning-for-phdi-lib.md

### PR DESCRIPTION
**Pull request by bamader**
This PR introduces an ADR for how we're going to do version numbers for PHDI going forward. The ADR itself is fairly self-explanatory.

### CODE REVIEW

File: docs/decisions/0009-versioning-for-phdi-lib.md [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+3. **Sequential Number Versioning**
+- Defined as versioning that begins with "1" and increments the count with each successive software push
+- Examples: Not really any that work out in the wild
+- Pros:
+  * Largely the same as date-of-release versioning without the drawback of an imposed / artificial schedule
+- Cons:
+  * Also susceptible to version verbosity, as with date versioning
+  * The optics of a massively large number for a version number (i.e. 427) could reduce confidence in our ability to make correct software the first time
+  * Creates the temptation to not increment the version number to avoid the above problem, which then eliminates consistency as well
+5. **Semantic Versioning (SemVer)**
```
Comment (KennethSkylight) [inline docs/decisions/0009-versioning-for-phdi-lib.md:49]: You skipped 4

**Review by KennethSkylight — CHANGES_REQUESTED** [review]
LGTM. Just one fix needed

File: docs/decisions/0009-versioning-for-phdi-lib.md (hunk 1 above)
Comment (bamader) [reply docs/decisions/0009-versioning-for-phdi-lib.md:49]: LOL wow dat automatic numbering doe

File: docs/decisions/0009-versioning-for-phdi-lib.md [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+
+Date: 2022-6-28
+
+## Status
+
+Accepted
+
+## Context and Problem Statement
+
+As the PHDI building blocks software continues to develop, documenting the versions that the software goes through will become increasingly important. Users will need to be able to track released versions of the software to ensure that they're using the correct version for their use cases, and so that they can receive proper support in the event they experience a software problem. Developers need to be able to document introductions of new features and bug fixes in existing features. Software versioning allows both groups to avoid confusion around the state of PHDI building blocks library.
```
Comment (bryan-skylight) [inline docs/decisions/0009-versioning-for-phdi-lib.md:11]: We should elaborate a little more here, pointing out the specific problems Semantic Versioning seems to want to account for, specifically around what the major, minor, and patch versions represent. For example, something like "Users need a mechanism by which to both ensure they're using the appropriate version of the code for their use case as well as identify when updating the library could break their current functionality." would help point out that the purpose behind the major version number is to identify when breaking changes have potentially been introduced.

File: docs/decisions/0009-versioning-for-phdi-lib.md [hunk 3]
```diff
@@ -0,0 +1,68 @@
+# 9. Formalize software version numbers using Semantic Versioning (SemVer)
```
Comment (bryan-skylight) [inline docs/decisions/0009-versioning-for-phdi-lib.md:1]: Can we shorten this to something like "Implement Semantic Versioning"? At the very least, I don't think the shortening of "Semantic Versioning" should be put in the title of the ADR.

File: docs/decisions/0009-versioning-for-phdi-lib.md [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+
+Accepted
+
+## Context and Problem Statement
+
+As the PHDI building blocks software continues to develop, documenting the versions that the software goes through will become increasingly important. Users will need to be able to track released versions of the software to ensure that they're using the correct version for their use cases, and so that they can receive proper support in the event they experience a software problem. Developers need to be able to document introductions of new features and bug fixes in existing features. Software versioning allows both groups to avoid confusion around the state of PHDI building blocks library.
+
+## Decision Drivers
+
+* **Transparency**: The adopted versioning should clearly convey relevant information about the software, such as the presence or absence of major features as well as the presence or absence of particular fixes.
```
[the remark this answers is not in the document]
Comment (bamader) [reply docs/decisions/0009-versioning-for-phdi-lib.md:15]: Yeah you're right. I don't know why I thought two decision drivers wasn't enough, but the primary points of using versioning are to make sure things run consistently, people can be comfortable knowing they're talking about the same version of the software, and that changes can be recognized along particular dimensions (i.e. major vs minor). I'm gonna dump transparency.

File: docs/decisions/0009-versioning-for-phdi-lib.md [hunk 5]
```diff
@@ -0,0 +1,68 @@
+# 9. Formalize software version numbers using Semantic Versioning (SemVer)
+
+Date: 2022-6-28
```
[the remark this answers is not in the document]
Comment (bamader) [reply docs/decisions/0009-versioning-for-phdi-lib.md:3]: Whoops missed the 0

File: docs/decisions/0009-versioning-for-phdi-lib.md [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+Accepted
+
+## Context and Problem Statement
+
+As the PHDI building blocks software continues to develop, documenting the versions that the software goes through will become increasingly important. Users will need to be able to track released versions of the software to ensure that they're using the correct version for their use cases, and so that they can receive proper support in the event they experience a software problem. Developers need to be able to document introductions of new features and bug fixes in existing features. Software versioning allows both groups to avoid confusion around the state of PHDI building blocks library.
+
+## Decision Drivers
+
+* **Transparency**: The adopted versioning should clearly convey relevant information about the software, such as the presence or absence of major features as well as the presence or absence of particular fixes.
+* **Consistency**: The adopted versioning should communicate that two identically-versioned instances of the software should behave identically; in other words, version X of the software on one machine should have the exact same characteristics as version X of the software on a different machine.
```
[the remark this answers is not in the document]
Comment (bamader) [reply docs/decisions/0009-versioning-for-phdi-lib.md:16]: After all, that's better than an ADR whose considered options are "Yes, implement versioning" and "No, continue into anarchy"

File: docs/decisions/0009-versioning-for-phdi-lib.md [hunk 7]
```diff
@@ -0,0 +1,68 @@
...
+* **Interpretability**: The adopted versioning should be easily interpretable to both a user and a developer. Users should be able to tell at a glance what version of the software they are using, and they should be able to understand the differences between two similarly-versioned instances of the software.
```
[the remark this answers is not in the document]
Comment (bamader) [reply docs/decisions/0009-versioning-for-phdi-lib.md:17]: That last sentence is a really good one and something I'm going to fold into interpretability. I think a user legitimately should be able to tell the magnitude of how much their functionality is going to change directly from a version number, and semantic versioning flat out gives us that.

File: docs/decisions/0009-versioning-for-phdi-lib.md [hunk 8]
```diff
@@ -0,0 +1,68 @@
...
+3. **Sequential Number Versioning**
+- Defined as versioning that begins with "1" and increments the count with each successive software push
+- Examples: Not really any that work out in the wild
+- Pros:
+  * Largely the same as date-of-release versioning without the drawback of an imposed / artificial schedule
+- Cons:
+  * Also susceptible to version verbosity, as with date versioning
+  * The optics of a massively large number for a version number (i.e. 427) could reduce confidence in our ability to make correct software the first time
+  * Creates the temptation to not increment the version number to avoid the above problem, which then eliminates consistency as well
+4. **Semantic Versioning (SemVer)**
```
[the remark this answers is not in the document]
Comment (bamader) [reply docs/decisions/0009-versioning-for-phdi-lib.md:49]: Lol sure I'll slash it out

**Comment by bamader** [comment]
Bryan, you gave me some good food for thought with those decision drivers. I decided to go down to just two but to blend the approach between justifying adopting versioning and justifying which versioning system to use. I think the two really go hand in hand, and it's not hard to talk about one without talking about the other. I think some of the points I was trying to make are now more clearly aligned with the standards we set out. See what you think.

File: docs/decisions/0009-versioning-for-phdi-lib.md [hunk 9]
```diff
@@ -0,0 +1,66 @@
...
+  * Well-documented, formally used process that reduces ambiguity on the developer side, to the extent that many major software developers around the world use it
+  * Optics around version numbers make intuitive sense to users; in other words, they should "feel natural" in a world where things like e.g. operating system numbers are of the form 14.12.21, which allows developers to create lots of changes and alter the version number without creating the feeling of an unstably changing software platform
+- Cons:
+  * May require careful consideration and discussion around whether a change has actually "broken" the existing API
+  * Bug fix / patch number can sometimes be treated as a "catch all" bucket and wind up spiraling to large numbers like sequential number versioning
+
+## Decision Outcome
+
+We are adopting Semantic Versioning to document the PHDI building blocks library's development. It is robust, standardized, and well-documented with precise protocols for when things change and what those changes should be. It will achieve all three major decision drivers and provide additional clarity for both users and developers.
```
Comment (bryan-skylight) [inline docs/decisions/0009-versioning-for-phdi-lib.md:62]: > It will achieve all three major decision drivers

This sentence needs to be updated now that we've gone down to two decision drivers.

**Review by bryan-skylight — APPROVED** [review]
LGTM with the exception of one comment. Fix that and I think we're good to land this.

**Review by KennethSkylight — APPROVED** [review]
